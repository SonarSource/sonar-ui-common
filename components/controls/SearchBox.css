/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.search-box {
  position: relative;
  display: inline-block;
  vertical-align: middle;
  font-size: 0;
  white-space: nowrap;
}

.search-box,
.search-box-input {
  width: 100%;
  max-width: 300px;
}

.search-box-input {
  /* for magnifier icon */
  padding-left: var(--controlHeight) !important;
  /* for clear button */
  padding-right: var(--controlHeight) !important;
  font-size: var(--baseFontSize);
}

.search-box-input::-webkit-search-decoration,
.search-box-input::-webkit-search-cancel-button,
.search-box-input::-webkit-search-results-button,
.search-box-input::-webkit-search-results-decoration {
  -webkit-appearance: none;
  display: none;
}

.search-box-input::-ms-clear,
.search-box-input::-ms-reveal {
  display: none;
  width: 0;
  height: 0;
}

.search-box-note {
  position: absolute;
  top: 1px;
  left: 40px;
  right: var(--controlHeight);
  line-height: var(--controlHeight);
  color: var(--secondFontColor);
  font-size: var(--smallFontSize);
  text-align: right;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  pointer-events: none;
}

.search-box-input:focus ~ .search-box-magnifier {
  color: var(--blue);
}

.search-box-magnifier {
  position: absolute;
  top: 4px;
  left: 4px;
  color: var(--gray60);
  transition: color 0.3s ease;
}

.search-box > .spinner {
  position: absolute;
  top: 4px;
  left: 5px;
}

.search-box-clear {
  position: absolute;
  top: 4px;
  right: 4px;
}

.search-box-input-note {
  position: absolute;
  top: 100%;
  left: 0;
  line-height: 1;
  color: var(--secondFontColor);
  font-size: var(--smallFontSize);
  white-space: nowrap;
}
