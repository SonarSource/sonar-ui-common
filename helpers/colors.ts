/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable no-bitwise, no-mixed-operators */
export function stringToColor(str: string) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  let color = '#';
  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += ('00' + value.toString(16)).substr(-2);
  }
  return color;
}

export function isDarkColor(color: string) {
  color = color.substr(1);
  if (color.length === 3) {
    // shortcut notation: #f90
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }
  const rgb = parseInt(color.substr(1), 16);
  const r = (rgb >> 16) & 0xff;
  const g = (rgb >> 8) & 0xff;
  const b = (rgb >> 0) & 0xff;
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  return luma < 140;
}

export function getTextColor(background: string, dark = '#222', light = '#fff') {
  return isDarkColor(background) ? light : dark;
}
