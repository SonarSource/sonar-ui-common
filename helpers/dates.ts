/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as parse from 'date-fns/parse';

function pad(number: number) {
  if (number < 10) {
    return '0' + number.toString();
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}
