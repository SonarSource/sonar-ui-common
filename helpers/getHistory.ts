/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { createHistory, History } from 'history';
import { useRouterHistory } from 'react-router';

let history: History;

function ensureHistory() {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  history = useRouterHistory(createHistory)({
    basename: (window as any).baseUrl,
  });
  return history;
}

export default function getHistory() {
  return history ? history : ensureHistory();
}
