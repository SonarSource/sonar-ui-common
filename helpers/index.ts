/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export * from './colors';
export * from './cookies';
export * from './csv';
export * from './dates';
export * from './getHistory';
export * from './handleRequiredAuthentication';
export * from './init';
export * from './keycodes';
export * from './l10n';
export * from './measures';
export * from './pages';
export * from './path';
export * from './query';
export * from './ratings';
export * from './request';
export * from './scrolling';
export * from './search';
export {
  get as getFromLocalStorage,
  remove as removeFromLocalStorage,
  save as saveInLocalStorage,
} from './storage';
export * from './strings';
export * from './testUtils';
export * from './urls';
