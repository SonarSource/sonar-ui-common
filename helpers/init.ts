/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import type { Messages } from './l10n';

let urlContext: string; // Is the base path (web context) in SQ
let messages: Messages | undefined;
let locale: string | undefined;

export const IS_SSR = typeof window === 'undefined';
export const DEFAULT_LOCALE = 'en';
export const DEFAULT_MESSAGES = {
  // eslint-disable-next-line camelcase
  default_error_message: 'The request cannot be processed. Try again later.',
};
const LOGGER_PREFIX = 'sonar-ui-common init:';

export default {
  setUrlContext(newUrlContext: string) {
    urlContext = newUrlContext;
    return this;
  },
  setLocale(newLocale: string) {
    locale = newLocale;
    return this;
  },
  setMessages(newMessages: Messages) {
    messages = newMessages;
    return this;
  },
};

export function getMessages() {
  if (typeof messages === 'undefined') {
    logWarning('L10n messages are not initialized. Use default messages.');
    return DEFAULT_MESSAGES;
  }
  return messages;
}

export function getLocale() {
  if (typeof locale === 'undefined') {
    logWarning('L10n locale is not initialized. Use default locale.');
    return DEFAULT_LOCALE;
  }
  return locale;
}

export function getUrlContext() {
  if (typeof urlContext === 'undefined') {
    throw new Error(
      `${LOGGER_PREFIX} web context needs to be initialized by Initializer.setUrlContext before being used`
    );
  }
  return urlContext;
}

function logWarning(message: string) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.warn(`${LOGGER_PREFIX} ${message}`);
  }
}
