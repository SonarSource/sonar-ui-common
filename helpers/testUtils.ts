/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export const KEYCODE_MAP: { [keycode: number]: string } = {
  13: 'enter',
  37: 'left',
  38: 'up',
  39: 'right',
  40: 'down'
};

export function keydown(key: number | string): void {
  let keyCode;
  if (typeof key === 'number') {
    keyCode = key;
  } else {
    // eslint-disable-next-line no-console
    console.warn('Using strings in keydown() is deprecated. Consider using the KeyCodes enum.');
    const mapped = Object.entries(KEYCODE_MAP).find(([_, value]) => value === key);
    if (!mapped) {
      throw new Error(`Cannot map key "${key}" to a keyCode!`);
    }
    keyCode = mapped[0];
  }

  const event = new KeyboardEvent('keydown', { keyCode, which: keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.documentElement.clientHeight/clientWidth` are getters by default,
  // so we need to redefine them. Pass `configurable: true` to allow to redefine
  // the properties multiple times.
  if (width) {
    Object.defineProperty(document.documentElement, 'clientWidth', {
      configurable: true,
      value: width
    });
  }
  if (height) {
    Object.defineProperty(document.documentElement, 'clientHeight', {
      configurable: true,
      value: height
    });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function scrollTo({ left = 0, top = 0 }) {
  Object.defineProperty(window, 'pageYOffset', { value: top });
  Object.defineProperty(window, 'pageXOffset', { value: left });
  const resizeEvent = new Event('scroll');
  window.dispatchEvent(resizeEvent);
}

export function setNodeRect({ width = 50, height = 50, left = 0, top = 0 }) {
  const { findDOMNode } = require('react-dom');
  const element = document.createElement('div');
  Object.defineProperty(element, 'getBoundingClientRect', {
    value: () => ({ width, height, left, top })
  });
  findDOMNode.mockReturnValue(element);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
