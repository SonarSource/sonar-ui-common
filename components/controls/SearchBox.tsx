/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import { Cancelable, debounce } from 'lodash';
import * as React from 'react';
import { translate, translateWithParameters } from '../../helpers/l10n';
import SearchIcon from '../icons/SearchIcon';
import DeferredSpinner from '../ui/DeferredSpinner';
import { ClearButton } from './buttons';
import './SearchBox.css';

interface Props {
  autoFocus?: boolean;
  className?: string;
  id?: string;
  innerRef?: (node: HTMLInputElement | null) => void;
  loading?: boolean;
  minLength?: number;
  onChange: (value: string) => void;
  onClick?: React.MouseEventHandler<HTMLInputElement>;
  onFocus?: React.FocusEventHandler<HTMLInputElement>;
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  placeholder: string;
  value?: string;
}

interface State {
  value: string;
}

export default class SearchBox extends React.PureComponent<Props, State> {
  debouncedOnChange: ((query: string) => void) & Cancelable;
  input?: HTMLInputElement | null;

  constructor(props: Props) {
    super(props);
    this.state = { value: props.value || '' };
    this.debouncedOnChange = debounce(this.props.onChange, 250);
  }

  componentDidUpdate(prevProps: Props) {
    if (
      // input is controlled
      this.props.value !== undefined &&
      // parent is aware of last change
      // can happen when previous value was less than min length
      this.state.value === prevProps.value &&
      this.state.value !== this.props.value
    ) {
      this.setState({ value: this.props.value });
    }
  }

  changeValue = (value: string, debounced = true) => {
    const { minLength } = this.props;
    if (value.length === 0) {
      // immediately notify when value is empty
      this.props.onChange('');
      // and cancel scheduled callback
      this.debouncedOnChange.cancel();
    } else if (!minLength || minLength <= value.length) {
      if (debounced) {
        this.debouncedOnChange(value);
      } else {
        this.props.onChange(value);
      }
    }
  };

  handleInputChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
    const { value } = event.currentTarget;
    this.setState({ value });
    this.changeValue(value);
  };

  handleInputKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.keyCode === 27) {
      // escape
      event.preventDefault();
      this.handleResetClick();
    }
    if (this.props.onKeyDown) {
      this.props.onKeyDown(event);
    }
  };

  handleResetClick = () => {
    this.changeValue('', false);
    if (this.props.value === undefined || this.props.value === '') {
      this.setState({ value: '' });
    }
    if (this.input) {
      this.input.focus();
    }
  };

  ref = (node: HTMLInputElement | null) => {
    this.input = node;
    if (this.props.innerRef) {
      this.props.innerRef(node);
    }
  };

  render() {
    const { loading, minLength } = this.props;
    const { value } = this.state;

    const inputClassName = classNames('search-box-input', {
      touched: value.length > 0 && (!minLength || minLength > value.length),
    });

    const tooShort = minLength !== undefined && value.length > 0 && value.length < minLength;

    return (
      <div
        className={classNames('search-box', this.props.className)}
        id={this.props.id}
        title={tooShort ? translateWithParameters('select2.tooShort', minLength!) : ''}>
        <input
          aria-label={translate('search_verb')}
          autoComplete="off"
          autoFocus={this.props.autoFocus}
          className={inputClassName}
          maxLength={100}
          onChange={this.handleInputChange}
          onClick={this.props.onClick}
          onFocus={this.props.onFocus}
          onKeyDown={this.handleInputKeyDown}
          placeholder={this.props.placeholder}
          ref={this.ref}
          type="search"
          value={value}
        />

        <DeferredSpinner loading={loading !== undefined ? loading : false}>
          <SearchIcon className="search-box-magnifier" />
        </DeferredSpinner>

        {value && (
          <ClearButton
            aria-label={translate('clear')}
            className="button-tiny search-box-clear"
            iconProps={{ size: 12 }}
            onClick={this.handleResetClick}
          />
        )}

        {tooShort && (
          <span className="search-box-note">
            {translateWithParameters('select2.tooShort', minLength!)}
          </span>
        )}
      </div>
    );
  }
}
