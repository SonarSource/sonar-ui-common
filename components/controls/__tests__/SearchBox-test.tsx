/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { change, click } from '../../../helpers/testUtils';
import SearchBox from '../SearchBox';

jest.mock('lodash', () => {
  const lodash = require.requireActual('lodash');
  const debounce = (fn: Function) => {
    const debounced: any = (...args: any[]) => fn(...args);
    debounced.cancel = jest.fn();
    return debounced;
  };
  return Object.assign({}, lodash, { debounce });
});

it('renders', () => {
  const wrapper = shallow(
    <SearchBox minLength={2} onChange={jest.fn()} placeholder="placeholder" value="foo" />
  );
  expect(wrapper).toMatchSnapshot();
});

it('warns when input is too short', () => {
  const wrapper = shallow(
    <SearchBox minLength={2} onChange={jest.fn()} placeholder="placeholder" value="f" />
  );
  expect(wrapper.find('.search-box-note').exists()).toBe(true);
});

it('shows clear button only when there is a value', () => {
  const wrapper = shallow(<SearchBox onChange={jest.fn()} placeholder="placeholder" value="f" />);
  expect(wrapper.find('.search-box-clear').exists()).toBe(true);
  wrapper.setProps({ value: '' });
  expect(wrapper.find('.search-box-clear').exists()).toBe(false);
});

it('attaches ref', () => {
  const ref = jest.fn();
  mount(<SearchBox innerRef={ref} onChange={jest.fn()} placeholder="placeholder" value="f" />);
  expect(ref).toBeCalled();
  expect(ref.mock.calls[0][0]).toBeInstanceOf(HTMLInputElement);
});

it('resets', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<SearchBox onChange={onChange} placeholder="placeholder" value="f" />);
  click(wrapper.find('.search-box-clear'));
  expect(onChange).toBeCalledWith('');
});

it('changes', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<SearchBox onChange={onChange} placeholder="placeholder" value="f" />);
  change(wrapper.find('.search-box-input'), 'foo');
  expect(onChange).toBeCalledWith('foo');
});

it('does not change when value is too short', () => {
  const onChange = jest.fn();
  const wrapper = shallow(
    <SearchBox minLength={3} onChange={onChange} placeholder="placeholder" value="" />
  );
  change(wrapper.find('.search-box-input'), 'fo');
  expect(onChange).not.toBeCalled();
});
