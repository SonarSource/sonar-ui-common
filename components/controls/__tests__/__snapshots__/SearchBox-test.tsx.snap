// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="search-box"
  title=""
>
  <input
    aria-label="search_verb"
    autoComplete="off"
    className="search-box-input"
    maxLength={100}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="placeholder"
    type="search"
    value="foo"
  />
  <DeferredSpinner
    loading={false}
  >
    <SearchIcon
      className="search-box-magnifier"
    />
  </DeferredSpinner>
  <ClearButton
    aria-label="clear"
    className="button-tiny search-box-clear"
    iconProps={
      Object {
        "size": 12,
      }
    }
    onClick={[Function]}
  />
</div>
`;
